dao-voting/tests/harness.rs

```diff
@@ -0,0 +1,897 @@
+            deployer
+                .dao_voting
+                .constructor(gov_token_id)
+                .call()
+                .await
+                .unwrap()
+                .value;
+
+            let call_data = get_call_data(gov_token_id);
+
+            deployer
+                .dao_voting
+                .create_proposal(10, 0, call_data)
+                .call()
+                .await
+                .unwrap()
+                .value;
+        }
+
+        #[tokio::test]
+        #[should_panic]
+        async fn panics_with_over_hundred_acceptance_percentage() {
+            let (_gov_token, gov_token_id, deployer, _user, _asset_amount) = setup().await;
+            deployer
+                .dao_voting
+                .constructor(gov_token_id)
+                .call()
+                .await
+                .unwrap()
+                .value;
+
+            let call_data = get_call_data(gov_token_id);
+
+            deployer
+                .dao_voting
+                .create_proposal(101, 10, call_data)
+                .call()
+                .await
+                .unwrap()
+                .value;
+        }
+    }
+}
+
+mod deposit {
+    use super::*;
+
+    mod success {
+        use super::*;
+
+        #[tokio::test]
+        async fn user_can_deposit() {
+            let (_gov_token, gov_token_id, deployer, user, asset_amount) = setup().await;
+
+            assert!(
+                deployer
+                    .gov_token
+                    .unwrap()
+                    .mint_and_send_to_address(100, user.wallet.address())
+                    .append_variable_outputs(1)
+                    .call()
+                    .await
+                    .unwrap()
+                    .value
+            );
+
+            deployer
+                .dao_voting
+                .constructor(gov_token_id)
+                .call()
+                .await
+                .unwrap()
+                .value;
+
+            assert_eq!(deployer.dao_voting.balance().call().await.unwrap().value, 0);
+
+            assert_eq!(
+                user.dao_voting
+                    .user_balance(Identity::Address(user.wallet.address()))
+                    .call()
+                    .await
+                    .unwrap()
+                    .value,
+                0
+            );
+
+            let tx_params = TxParameters::new(None, Some(1_000_000), None, None);
+            let call_params = CallParameters::new(
+                Some(asset_amount),
+                Some(AssetId::from(*gov_token_id)),
+                Some(100_000),
+            );
+            user.dao_voting
+                .deposit()
+                .tx_params(tx_params)
+                .call_params(call_params)
+                .call()
+                .await
+                .unwrap()
+                .value;
+
+            assert_eq!(
+                deployer.dao_voting.balance().call().await.unwrap().value,
+                asset_amount
```

Comment:
Note that it is possible to make this test pass while having the contract not actually take a deposit because you are not making sure that the asset_amount != 0. Passing in params into functions instead of hardcoding is a good practice

Reply:
hmm I don't think its possible for that to happen.  The contract's deposit function requires the amount to be greater than 0.  Therefore the function will fail if `asset_amount` is 0.